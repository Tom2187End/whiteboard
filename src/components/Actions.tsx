import React from "react";
import { AppState } from "../types";
import { ExcalidrawElement } from "../element/types";
import { ActionManager } from "../actions/manager";
import { hasBackground, hasStroke, hasText, getTargetElement } from "../scene";
import { t } from "../i18n";
import { SHAPES } from "../shapes";
import { ToolButton } from "./ToolButton";
import { capitalizeString, setCursorForShape } from "../utils";
import Stack from "./Stack";
import useIsMobile from "../is-mobile";

export function SelectedShapeActions({
  appState,
  elements,
  renderAction,
  elementType,
}: {
  appState: AppState;
  elements: readonly ExcalidrawElement[];
  renderAction: ActionManager["renderAction"];
  elementType: ExcalidrawElement["type"];
}) {
  const targetElements = getTargetElement(elements, appState);
  const isEditing = Boolean(appState.editingElement);
  const isMobile = useIsMobile();

  return (
    <div className="panelColumn">
      {renderAction("changeStrokeColor")}
      {(hasBackground(elementType) ||
        targetElements.some((element) => hasBackground(element.type))) && (
        <>
          {renderAction("changeBackgroundColor")}

          {renderAction("changeFillStyle")}
        </>
      )}

      {(hasStroke(elementType) ||
        targetElements.some((element) => hasStroke(element.type))) && (
        <>
          {renderAction("changeStrokeWidth")}

          {renderAction("changeSloppiness")}
        </>
      )}

      {(hasText(elementType) ||
        targetElements.some((element) => hasText(element.type))) && (
        <>
          {renderAction("changeFontSize")}

          {renderAction("changeFontFamily")}
        </>
      )}

      {renderAction("changeOpacity")}

      <fieldset>
        <legend>{t("labels.layers")}</legend>
        <div className="buttonList">
          {renderAction("sendToBack")}
          {renderAction("sendBackward")}
          {renderAction("bringToFront")}
          {renderAction("bringForward")}
        </div>
      </fieldset>
      {!isMobile && !isEditing && targetElements.length > 0 && (
        <fieldset>
          <legend>{t("labels.actions")}</legend>
          <div className="buttonList">
            {renderAction("duplicateSelection")}
            {renderAction("deleteSelectedElements")}
          </div>
        </fieldset>
      )}
    </div>
  );
}

export function ShapesSwitcher({
  elementType,
  setAppState,
  setElements,
  elements,
}: {
  elementType: ExcalidrawElement["type"];
  setAppState: any;
  setElements: any;
  elements: readonly ExcalidrawElement[];
}) {
  return (
    <>
      {SHAPES.map(({ value, icon }, index) => {
        const label = t(`toolBar.${value}`);
        const shortcut = `${capitalizeString(value)[0]} ${t(
          "shortcutsDialog.or",
        )} ${index + 1}`;
        return (
          <ToolButton
            key={value}
            type="radio"
            icon={icon}
            checked={elementType === value}
            name="editor-current-shape"
            title={`${capitalizeString(label)} â€” ${shortcut}`}
            keyBindingLabel={`${index + 1}`}
            aria-label={capitalizeString(label)}
            aria-keyshortcuts={`${label[0]} ${index + 1}`}
            data-testid={value}
            onChange={() => {
              setAppState({
                elementType: value,
                multiElement: null,
                selectedElementIds: {},
              });
              setCursorForShape(value);
              setAppState({});
            }}
          ></ToolButton>
        );
      })}
    </>
  );
}

export function ZoomActions({
  renderAction,
  zoom,
}: {
  renderAction: ActionManager["renderAction"];
  zoom: number;
}) {
  return (
    <Stack.Col gap={1}>
      <Stack.Row gap={1} align="center">
        {renderAction("zoomIn")}
        {renderAction("zoomOut")}
        {renderAction("resetZoom")}
        <div style={{ marginInlineStart: 4 }}>{(zoom * 100).toFixed(0)}%</div>
      </Stack.Row>
    </Stack.Col>
  );
}
