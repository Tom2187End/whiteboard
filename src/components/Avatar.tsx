import "./Avatar.scss";

import React from "react";

type AvatarProps = {
  children: string;
  onClick: (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => void;
  color: string;
};

export const Avatar = ({ children, color, onClick }: AvatarProps) => (
  <div className="Avatar" style={{ background: color }} onClick={onClick}>
    {children}
  </div>
);
