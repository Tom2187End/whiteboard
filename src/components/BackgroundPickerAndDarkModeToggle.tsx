import React from "react";
import { ActionManager } from "../actions/manager";
import { AppState } from "../types";
import { DarkModeToggle } from "./DarkModeToggle";

export const BackgroundPickerAndDarkModeToggle = ({
  appState,
  setAppState,
  actionManager,
}: {
  actionManager: ActionManager;
  appState: AppState;
  setAppState: any;
}) => (
  <div style={{ display: "flex" }}>
    {actionManager.renderAction("changeViewBackgroundColor")}
    <div style={{ marginInlineStart: "0.25rem" }}>
      <DarkModeToggle
        value={appState.appearance}
        onChange={(appearance) => {
          setAppState({ appearance });
        }}
      />
    </div>
  </div>
);
