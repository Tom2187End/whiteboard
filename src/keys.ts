export const isDarwin = /Mac|iPod|iPhone|iPad/.test(window.navigator.platform);

export const KEYS = {
  ARROW_LEFT: "ArrowLeft",
  ARROW_RIGHT: "ArrowRight",
  ARROW_DOWN: "ArrowDown",
  ARROW_UP: "ArrowUp",
  ENTER: "Enter",
  ESCAPE: "Escape",
  DELETE: "Delete",
  BACKSPACE: "Backspace",
  CTRL_OR_CMD: isDarwin ? "metaKey" : "ctrlKey",
  TAB: "Tab",
  SPACE: " ",
  QUESTION_MARK: "?",
  F_KEY_CODE: 70,
  ALT_KEY_CODE: 18,
  Z_KEY_CODE: 90,
  GRID_KEY_CODE: 222,
  G_KEY_CODE: 71,
  C_KEY_CODE: 67,
  V_KEY_CODE: 86,
} as const;

export type Key = keyof typeof KEYS;

export const isArrowKey = (keyCode: string) =>
  keyCode === KEYS.ARROW_LEFT ||
  keyCode === KEYS.ARROW_RIGHT ||
  keyCode === KEYS.ARROW_DOWN ||
  keyCode === KEYS.ARROW_UP;

export const getResizeCenterPointKey = (event: MouseEvent | KeyboardEvent) =>
  event.altKey || event.which === KEYS.ALT_KEY_CODE;

export const getResizeWithSidesSameLengthKey = (event: MouseEvent) =>
  event.shiftKey;

export const getRotateWithDiscreteAngleKey = (event: MouseEvent) =>
  event.shiftKey;
