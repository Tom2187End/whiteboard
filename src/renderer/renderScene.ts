import { RoughCanvas } from "roughjs/bin/canvas";
import { RoughSVG } from "roughjs/bin/svg";

import { FlooredNumber, AppState } from "../types";
import { ExcalidrawElement } from "../element/types";
import {
  getElementAbsoluteCoords,
  OMIT_SIDES_FOR_MULTIPLE_ELEMENTS,
  handlerRectanglesFromCoords,
  handlerRectangles,
  getCommonBounds,
  canResizeMutlipleElements,
} from "../element";

import { roundRect } from "./roundRect";
import { SceneState } from "../scene/types";
import {
  getScrollBars,
  SCROLLBAR_COLOR,
  SCROLLBAR_WIDTH,
} from "../scene/scrollbars";
import { getSelectedElements } from "../scene/selection";

import { renderElement, renderElementToSvg } from "./renderElement";
import colors from "../colors";

type HandlerRectanglesRet = keyof ReturnType<typeof handlerRectangles>;

function colorsForClientId(clientId: string) {
  // Naive way of getting an integer out of the clientId
  const sum = clientId.split("").reduce((a, str) => a + str.charCodeAt(0), 0);

  // Skip transparent background.
  const backgrounds = colors.elementBackground.slice(1);
  const strokes = colors.elementStroke.slice(1);
  return {
    background: backgrounds[sum % backgrounds.length],
    stroke: strokes[sum % strokes.length],
  };
}

function strokeRectWithRotation(
  context: CanvasRenderingContext2D,
  x: number,
  y: number,
  width: number,
  height: number,
  cx: number,
  cy: number,
  angle: number,
  fill?: boolean,
) {
  context.translate(cx, cy);
  context.rotate(angle);
  if (fill) {
    context.fillRect(x - cx, y - cy, width, height);
  }
  context.strokeRect(x - cx, y - cy, width, height);
  context.rotate(-angle);
  context.translate(-cx, -cy);
}

function strokeCircle(
  context: CanvasRenderingContext2D,
  x: number,
  y: number,
  width: number,
  height: number,
) {
  context.beginPath();
  context.arc(x + width / 2, y + height / 2, width / 2, 0, Math.PI * 2);
  context.fill();
  context.stroke();
}

export function renderScene(
  allElements: readonly ExcalidrawElement[],
  appState: AppState,
  selectionElement: ExcalidrawElement | null,
  scale: number,
  rc: RoughCanvas,
  canvas: HTMLCanvasElement,
  sceneState: SceneState,
  // extra options, currently passed by export helper
  {
    renderScrollbars = true,
    renderSelection = true,
    // Whether to employ render optimizations to improve performance.
    // Should not be turned on for export operations and similar, because it
    //  doesn't guarantee pixel-perfect output.
    renderOptimizations = false,
  }: {
    renderScrollbars?: boolean;
    renderSelection?: boolean;
    renderOptimizations?: boolean;
  } = {},
) {
  if (!canvas) {
    return { atLeastOneVisibleElement: false };
  }

  const elements = allElements.filter((element) => !element.isDeleted);

  const context = canvas.getContext("2d")!;
  context.scale(scale, scale);

  // When doing calculations based on canvas width we should used normalized one
  const normalizedCanvasWidth = canvas.width / scale;
  const normalizedCanvasHeight = canvas.height / scale;

  // Paint background
  if (typeof sceneState.viewBackgroundColor === "string") {
    const hasTransparence =
      sceneState.viewBackgroundColor === "transparent" ||
      sceneState.viewBackgroundColor.length === 5 ||
      sceneState.viewBackgroundColor.length === 9;
    if (hasTransparence) {
      context.clearRect(0, 0, normalizedCanvasWidth, normalizedCanvasHeight);
    }
    const fillStyle = context.fillStyle;
    context.fillStyle = sceneState.viewBackgroundColor;
    context.fillRect(0, 0, normalizedCanvasWidth, normalizedCanvasHeight);
    context.fillStyle = fillStyle;
  } else {
    context.clearRect(0, 0, normalizedCanvasWidth, normalizedCanvasHeight);
  }

  // Apply zoom
  const zoomTranslationX = (-normalizedCanvasWidth * (sceneState.zoom - 1)) / 2;
  const zoomTranslationY =
    (-normalizedCanvasHeight * (sceneState.zoom - 1)) / 2;
  context.translate(zoomTranslationX, zoomTranslationY);
  context.scale(sceneState.zoom, sceneState.zoom);

  // Paint visible elements
  const visibleElements = elements.filter((element) =>
    isVisibleElement(
      element,
      normalizedCanvasWidth,
      normalizedCanvasHeight,
      sceneState,
    ),
  );

  visibleElements.forEach((element) => {
    renderElement(element, rc, context, renderOptimizations, sceneState);
  });

  // Pain selection element
  if (selectionElement) {
    renderElement(
      selectionElement,
      rc,
      context,
      renderOptimizations,
      sceneState,
    );
  }

  // Paint selected elements
  if (renderSelection) {
    context.translate(sceneState.scrollX, sceneState.scrollY);

    const selections = elements.reduce((acc, element) => {
      const selectionColors = [];
      // local user
      if (appState.selectedElementIds[element.id]) {
        selectionColors.push("#000000");
      }
      // remote users
      if (sceneState.remoteSelectedElementIds[element.id]) {
        selectionColors.push(
          ...sceneState.remoteSelectedElementIds[element.id].map((socketId) => {
            const { background } = colorsForClientId(socketId);
            return background;
          }),
        );
      }
      if (selectionColors.length) {
        acc.push({ element, selectionColors });
      }
      return acc;
    }, [] as { element: ExcalidrawElement; selectionColors: string[] }[]);

    selections.forEach(({ element, selectionColors }) => {
      const [
        elementX1,
        elementY1,
        elementX2,
        elementY2,
      ] = getElementAbsoluteCoords(element);

      const elementWidth = elementX2 - elementX1;
      const elementHeight = elementY2 - elementY1;

      const initialLineDash = context.getLineDash();
      const lineWidth = context.lineWidth;
      const lineDashOffset = context.lineDashOffset;
      const strokeStyle = context.strokeStyle;

      const dashedLinePadding = 4 / sceneState.zoom;
      const dashWidth = 8 / sceneState.zoom;
      const spaceWidth = 4 / sceneState.zoom;

      context.lineWidth = 1 / sceneState.zoom;

      const count = selectionColors.length;
      for (var i = 0; i < count; ++i) {
        context.strokeStyle = selectionColors[i];
        context.setLineDash([
          dashWidth,
          spaceWidth + (dashWidth + spaceWidth) * (count - 1),
        ]);
        context.lineDashOffset = (dashWidth + spaceWidth) * i;
        strokeRectWithRotation(
          context,
          elementX1 - dashedLinePadding,
          elementY1 - dashedLinePadding,
          elementWidth + dashedLinePadding * 2,
          elementHeight + dashedLinePadding * 2,
          elementX1 + elementWidth / 2,
          elementY1 + elementHeight / 2,
          element.angle,
        );
      }
      context.lineDashOffset = lineDashOffset;
      context.strokeStyle = strokeStyle;
      context.lineWidth = lineWidth;
      context.setLineDash(initialLineDash);
    });
    context.translate(-sceneState.scrollX, -sceneState.scrollY);

    const locallySelectedElements = getSelectedElements(elements, appState);

    // Paint resize handlers
    if (locallySelectedElements.length === 1) {
      context.translate(sceneState.scrollX, sceneState.scrollY);
      context.fillStyle = "#fff";
      const handlers = handlerRectangles(
        locallySelectedElements[0],
        sceneState.zoom,
      );
      Object.keys(handlers).forEach((key) => {
        const handler = handlers[key as HandlerRectanglesRet];
        if (handler !== undefined) {
          const lineWidth = context.lineWidth;
          context.lineWidth = 1 / sceneState.zoom;
          if (key === "rotation") {
            strokeCircle(
              context,
              handler[0],
              handler[1],
              handler[2],
              handler[3],
            );
          } else if (locallySelectedElements[0].type !== "text") {
            strokeRectWithRotation(
              context,
              handler[0],
              handler[1],
              handler[2],
              handler[3],
              handler[0] + handler[2] / 2,
              handler[1] + handler[3] / 2,
              locallySelectedElements[0].angle,
              true, // fill before stroke
            );
          }
          context.lineWidth = lineWidth;
        }
      });
      context.translate(-sceneState.scrollX, -sceneState.scrollY);
    } else if (locallySelectedElements.length > 1) {
      if (canResizeMutlipleElements(locallySelectedElements)) {
        const dashedLinePadding = 4 / sceneState.zoom;
        context.translate(sceneState.scrollX, sceneState.scrollY);
        context.fillStyle = "#fff";
        const [x1, y1, x2, y2] = getCommonBounds(locallySelectedElements);
        const initialLineDash = context.getLineDash();
        context.setLineDash([2 / sceneState.zoom]);
        const lineWidth = context.lineWidth;
        context.lineWidth = 1 / sceneState.zoom;
        strokeRectWithRotation(
          context,
          x1 - dashedLinePadding,
          y1 - dashedLinePadding,
          x2 - x1 + dashedLinePadding * 2,
          y2 - y1 + dashedLinePadding * 2,
          (x1 + x2) / 2,
          (y1 + y2) / 2,
          0,
        );
        context.lineWidth = lineWidth;
        context.setLineDash(initialLineDash);
        const handlers = handlerRectanglesFromCoords(
          [x1, y1, x2, y2],
          0,
          sceneState.zoom,
          undefined,
          OMIT_SIDES_FOR_MULTIPLE_ELEMENTS,
        );
        Object.keys(handlers).forEach((key) => {
          const handler = handlers[key as HandlerRectanglesRet];
          if (handler !== undefined) {
            const lineWidth = context.lineWidth;
            context.lineWidth = 1 / sceneState.zoom;
            strokeRectWithRotation(
              context,
              handler[0],
              handler[1],
              handler[2],
              handler[3],
              handler[0] + handler[2] / 2,
              handler[1] + handler[3] / 2,
              0,
              true, // fill before stroke
            );
            context.lineWidth = lineWidth;
          }
        });
        context.translate(-sceneState.scrollX, -sceneState.scrollY);
      }
    }
  }

  // Reset zoom
  context.scale(1 / sceneState.zoom, 1 / sceneState.zoom);
  context.translate(-zoomTranslationX, -zoomTranslationY);

  // Paint remote pointers
  for (const clientId in sceneState.remotePointerViewportCoords) {
    let { x, y } = sceneState.remotePointerViewportCoords[clientId];
    const username = sceneState.remotePointerUsernames[clientId];

    const width = 9;
    const height = 14;

    const isOutOfBounds =
      x < 0 ||
      x > normalizedCanvasWidth - width ||
      y < 0 ||
      y > normalizedCanvasHeight - height;

    x = Math.max(x, 0);
    x = Math.min(x, normalizedCanvasWidth - width);
    y = Math.max(y, 0);
    y = Math.min(y, normalizedCanvasHeight - height);

    const { background, stroke } = colorsForClientId(clientId);

    const strokeStyle = context.strokeStyle;
    const fillStyle = context.fillStyle;
    const globalAlpha = context.globalAlpha;
    context.strokeStyle = stroke;
    context.fillStyle = background;
    if (isOutOfBounds) {
      context.globalAlpha = 0.2;
    }

    if (
      sceneState.remotePointerButton &&
      sceneState.remotePointerButton[clientId] === "down"
    ) {
      context.beginPath();
      context.arc(x, y, 15, 0, 2 * Math.PI, false);
      context.lineWidth = 3;
      context.strokeStyle = "#ffffff88";
      context.stroke();
      context.closePath();

      context.beginPath();
      context.arc(x, y, 15, 0, 2 * Math.PI, false);
      context.lineWidth = 1;
      context.strokeStyle = stroke;
      context.stroke();
      context.closePath();
    }

    context.beginPath();
    context.moveTo(x, y);
    context.lineTo(x + 1, y + 14);
    context.lineTo(x + 4, y + 9);
    context.lineTo(x + 9, y + 10);
    context.lineTo(x, y);
    context.fill();
    context.stroke();

    if (!isOutOfBounds && username) {
      const offsetX = x + width;
      const offsetY = y + height;
      const paddingHorizontal = 4;
      const paddingVertical = 4;
      const measure = context.measureText(username);
      const measureHeight =
        measure.actualBoundingBoxDescent + measure.actualBoundingBoxAscent;

      // Border
      context.fillStyle = stroke;
      context.globalAlpha = globalAlpha;
      context.fillRect(
        offsetX - 1,
        offsetY - 1,
        measure.width + 2 * paddingHorizontal + 2,
        measureHeight + 2 * paddingVertical + 2,
      );
      // Background
      context.fillStyle = background;
      context.fillRect(
        offsetX,
        offsetY,
        measure.width + 2 * paddingHorizontal,
        measureHeight + 2 * paddingVertical,
      );
      context.fillStyle = "#ffffff";
      context.fillText(
        username,
        offsetX + paddingHorizontal,
        offsetY + paddingVertical + measure.actualBoundingBoxAscent,
      );
    }

    context.strokeStyle = strokeStyle;
    context.fillStyle = fillStyle;
    context.globalAlpha = globalAlpha;
    context.closePath();
  }

  // Paint scrollbars
  let scrollBars;
  if (renderScrollbars) {
    scrollBars = getScrollBars(
      elements,
      normalizedCanvasWidth,
      normalizedCanvasHeight,
      sceneState,
    );

    const fillStyle = context.fillStyle;
    const strokeStyle = context.strokeStyle;
    context.fillStyle = SCROLLBAR_COLOR;
    context.strokeStyle = "rgba(255,255,255,0.8)";
    [scrollBars.horizontal, scrollBars.vertical].forEach((scrollBar) => {
      if (scrollBar) {
        roundRect(
          context,
          scrollBar.x,
          scrollBar.y,
          scrollBar.width,
          scrollBar.height,
          SCROLLBAR_WIDTH / 2,
        );
      }
    });
    context.fillStyle = fillStyle;
    context.strokeStyle = strokeStyle;
  }

  context.scale(1 / scale, 1 / scale);

  return { atLeastOneVisibleElement: visibleElements.length > 0, scrollBars };
}

function isVisibleElement(
  element: ExcalidrawElement,
  viewportWidth: number,
  viewportHeight: number,
  {
    scrollX,
    scrollY,
    zoom,
  }: {
    scrollX: FlooredNumber;
    scrollY: FlooredNumber;
    zoom: number;
  },
) {
  const [x1, y1, x2, y2] = getElementAbsoluteCoords(element);

  // Apply zoom
  const viewportWidthWithZoom = viewportWidth / zoom;
  const viewportHeightWithZoom = viewportHeight / zoom;

  const viewportWidthDiff = viewportWidth - viewportWidthWithZoom;
  const viewportHeightDiff = viewportHeight - viewportHeightWithZoom;

  return (
    x2 + scrollX - viewportWidthDiff / 2 >= 0 &&
    x1 + scrollX - viewportWidthDiff / 2 <= viewportWidthWithZoom &&
    y2 + scrollY - viewportHeightDiff / 2 >= 0 &&
    y1 + scrollY - viewportHeightDiff / 2 <= viewportHeightWithZoom
  );
}

// This should be only called for exporting purposes
export function renderSceneToSvg(
  elements: readonly ExcalidrawElement[],
  rsvg: RoughSVG,
  svgRoot: SVGElement,
  {
    offsetX = 0,
    offsetY = 0,
  }: {
    offsetX?: number;
    offsetY?: number;
  } = {},
) {
  if (!svgRoot) {
    return;
  }
  // render elements
  elements.forEach((element) => {
    if (!element.isDeleted) {
      renderElementToSvg(
        element,
        rsvg,
        svgRoot,
        element.x + offsetX,
        element.y + offsetY,
      );
    }
  });
}
