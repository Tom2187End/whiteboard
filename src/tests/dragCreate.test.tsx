import React from "react";
import ReactDOM from "react-dom";
import { App } from "../components/App";
import * as Renderer from "../renderer/renderScene";
import { KEYS } from "../keys";
import { render, fireEvent } from "./test-utils";

// Unmount ReactDOM from root
ReactDOM.unmountComponentAtNode(document.getElementById("root")!);

const renderScene = jest.spyOn(Renderer, "renderScene");
beforeEach(() => {
  localStorage.clear();
  renderScene.mockClear();
});

const { __TEST__: h } = window;

describe("add element to the scene when pointer dragging long enough", () => {
  it("rectangle", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("rectangle");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;

    // start from (30, 20)
    fireEvent.pointerDown(canvas, { clientX: 30, clientY: 20 });

    // move to (60,70)
    fireEvent.pointerMove(canvas, { clientX: 60, clientY: 70 });

    // finish (position does not matter)
    fireEvent.pointerUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(4);
    expect(h.appState.selectionElement).toBeNull();

    expect(h.elements.length).toEqual(1);
    expect(h.elements[0].type).toEqual("rectangle");
    expect(h.elements[0].x).toEqual(30);
    expect(h.elements[0].y).toEqual(20);
    expect(h.elements[0].width).toEqual(30); // 60 - 30
    expect(h.elements[0].height).toEqual(50); // 70 - 20
  });

  it("ellipse", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("ellipse");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;

    // start from (30, 20)
    fireEvent.pointerDown(canvas, { clientX: 30, clientY: 20 });

    // move to (60,70)
    fireEvent.pointerMove(canvas, { clientX: 60, clientY: 70 });

    // finish (position does not matter)
    fireEvent.pointerUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(4);
    expect(h.appState.selectionElement).toBeNull();

    expect(h.elements.length).toEqual(1);
    expect(h.elements[0].type).toEqual("ellipse");
    expect(h.elements[0].x).toEqual(30);
    expect(h.elements[0].y).toEqual(20);
    expect(h.elements[0].width).toEqual(30); // 60 - 30
    expect(h.elements[0].height).toEqual(50); // 70 - 20
  });

  it("diamond", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("diamond");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;

    // start from (30, 20)
    fireEvent.pointerDown(canvas, { clientX: 30, clientY: 20 });

    // move to (60,70)
    fireEvent.pointerMove(canvas, { clientX: 60, clientY: 70 });

    // finish (position does not matter)
    fireEvent.pointerUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(4);
    expect(h.appState.selectionElement).toBeNull();

    expect(h.elements.length).toEqual(1);
    expect(h.elements[0].type).toEqual("diamond");
    expect(h.elements[0].x).toEqual(30);
    expect(h.elements[0].y).toEqual(20);
    expect(h.elements[0].width).toEqual(30); // 60 - 30
    expect(h.elements[0].height).toEqual(50); // 70 - 20
  });

  it("arrow", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("arrow");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;

    // start from (30, 20)
    fireEvent.pointerDown(canvas, { clientX: 30, clientY: 20 });

    // move to (60,70)
    fireEvent.pointerMove(canvas, { clientX: 60, clientY: 70 });

    // finish (position does not matter)
    fireEvent.pointerUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(4);
    expect(h.appState.selectionElement).toBeNull();

    expect(h.elements.length).toEqual(1);
    expect(h.elements[0].type).toEqual("arrow");
    expect(h.elements[0].x).toEqual(30);
    expect(h.elements[0].y).toEqual(20);
    expect(h.elements[0].points.length).toEqual(2);
    expect(h.elements[0].points[0]).toEqual([0, 0]);
    expect(h.elements[0].points[1]).toEqual([30, 50]); // (60 - 30, 70 - 20)
  });

  it("line", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("line");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;

    // start from (30, 20)
    fireEvent.pointerDown(canvas, { clientX: 30, clientY: 20 });

    // move to (60,70)
    fireEvent.pointerMove(canvas, { clientX: 60, clientY: 70 });

    // finish (position does not matter)
    fireEvent.pointerUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(4);
    expect(h.appState.selectionElement).toBeNull();

    expect(h.elements.length).toEqual(1);
    expect(h.elements[0].type).toEqual("line");
    expect(h.elements[0].x).toEqual(30);
    expect(h.elements[0].y).toEqual(20);
    expect(h.elements[0].points.length).toEqual(2);
    expect(h.elements[0].points[0]).toEqual([0, 0]);
    expect(h.elements[0].points[1]).toEqual([30, 50]); // (60 - 30, 70 - 20)
  });
});

describe("do not add element to the scene if size is too small", () => {
  it("rectangle", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("rectangle");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;

    // start from (30, 20)
    fireEvent.pointerDown(canvas, { clientX: 30, clientY: 20 });

    // finish (position does not matter)
    fireEvent.pointerUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(3);
    expect(h.appState.selectionElement).toBeNull();
    expect(h.elements.length).toEqual(0);
  });

  it("ellipse", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("ellipse");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;

    // start from (30, 20)
    fireEvent.pointerDown(canvas, { clientX: 30, clientY: 20 });

    // finish (position does not matter)
    fireEvent.pointerUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(3);
    expect(h.appState.selectionElement).toBeNull();
    expect(h.elements.length).toEqual(0);
  });

  it("diamond", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("diamond");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;

    // start from (30, 20)
    fireEvent.pointerDown(canvas, { clientX: 30, clientY: 20 });

    // finish (position does not matter)
    fireEvent.pointerUp(canvas);

    expect(renderScene).toHaveBeenCalledTimes(3);
    expect(h.appState.selectionElement).toBeNull();
    expect(h.elements.length).toEqual(0);
  });

  it("arrow", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("arrow");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;

    // start from (30, 20)
    fireEvent.pointerDown(canvas, { clientX: 30, clientY: 20 });

    // finish (position does not matter)
    fireEvent.pointerUp(canvas);

    // we need to finalize it because arrows and lines enter multi-mode
    fireEvent.keyDown(document, { key: KEYS.ENTER });

    expect(renderScene).toHaveBeenCalledTimes(4);
    expect(h.appState.selectionElement).toBeNull();
    expect(h.elements.length).toEqual(0);
  });

  it("line", () => {
    const { getByToolName, container } = render(<App />);
    // select tool
    const tool = getByToolName("line");
    fireEvent.click(tool);

    const canvas = container.querySelector("canvas")!;

    // start from (30, 20)
    fireEvent.pointerDown(canvas, { clientX: 30, clientY: 20 });

    // finish (position does not matter)
    fireEvent.pointerUp(canvas);

    // we need to finalize it because arrows and lines enter multi-mode
    fireEvent.keyDown(document, { key: KEYS.ENTER });

    expect(renderScene).toHaveBeenCalledTimes(4);
    expect(h.appState.selectionElement).toBeNull();
    expect(h.elements.length).toEqual(0);
  });
});
