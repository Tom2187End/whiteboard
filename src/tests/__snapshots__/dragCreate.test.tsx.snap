// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add element to the scene when pointer dragging long enough arrow 1`] = `1`;

exports[`add element to the scene when pointer dragging long enough arrow 2`] = `
Object {
  "angle": 0,
  "backgroundColor": "transparent",
  "fillStyle": "hachure",
  "groupIds": Array [],
  "height": 50,
  "id": "id0",
  "isDeleted": false,
  "lastCommittedPoint": null,
  "opacity": 100,
  "points": Array [
    Array [
      0,
      0,
    ],
    Array [
      30,
      50,
    ],
  ],
  "roughness": 1,
  "seed": 337897,
  "strokeColor": "#000000",
  "strokeStyle": "solid",
  "strokeWidth": 1,
  "type": "arrow",
  "version": 3,
  "versionNonce": 449462985,
  "width": 30,
  "x": 30,
  "y": 20,
}
`;

exports[`add element to the scene when pointer dragging long enough diamond 1`] = `1`;

exports[`add element to the scene when pointer dragging long enough diamond 2`] = `
Object {
  "angle": 0,
  "backgroundColor": "transparent",
  "fillStyle": "hachure",
  "groupIds": Array [],
  "height": 50,
  "id": "id0",
  "isDeleted": false,
  "opacity": 100,
  "roughness": 1,
  "seed": 337897,
  "strokeColor": "#000000",
  "strokeStyle": "solid",
  "strokeWidth": 1,
  "type": "diamond",
  "version": 2,
  "versionNonce": 1278240551,
  "width": 30,
  "x": 30,
  "y": 20,
}
`;

exports[`add element to the scene when pointer dragging long enough ellipse 1`] = `1`;

exports[`add element to the scene when pointer dragging long enough ellipse 2`] = `
Object {
  "angle": 0,
  "backgroundColor": "transparent",
  "fillStyle": "hachure",
  "groupIds": Array [],
  "height": 50,
  "id": "id0",
  "isDeleted": false,
  "opacity": 100,
  "roughness": 1,
  "seed": 337897,
  "strokeColor": "#000000",
  "strokeStyle": "solid",
  "strokeWidth": 1,
  "type": "ellipse",
  "version": 2,
  "versionNonce": 1278240551,
  "width": 30,
  "x": 30,
  "y": 20,
}
`;

exports[`add element to the scene when pointer dragging long enough line 1`] = `
Object {
  "angle": 0,
  "backgroundColor": "transparent",
  "fillStyle": "hachure",
  "groupIds": Array [],
  "height": 50,
  "id": "id0",
  "isDeleted": false,
  "lastCommittedPoint": null,
  "opacity": 100,
  "points": Array [
    Array [
      0,
      0,
    ],
    Array [
      30,
      50,
    ],
  ],
  "roughness": 1,
  "seed": 337897,
  "strokeColor": "#000000",
  "strokeStyle": "solid",
  "strokeWidth": 1,
  "type": "line",
  "version": 3,
  "versionNonce": 449462985,
  "width": 30,
  "x": 30,
  "y": 20,
}
`;

exports[`add element to the scene when pointer dragging long enough rectangle 1`] = `1`;

exports[`add element to the scene when pointer dragging long enough rectangle 2`] = `
Object {
  "angle": 0,
  "backgroundColor": "transparent",
  "fillStyle": "hachure",
  "groupIds": Array [],
  "height": 50,
  "id": "id0",
  "isDeleted": false,
  "opacity": 100,
  "roughness": 1,
  "seed": 337897,
  "strokeColor": "#000000",
  "strokeStyle": "solid",
  "strokeWidth": 1,
  "type": "rectangle",
  "version": 2,
  "versionNonce": 1278240551,
  "width": 30,
  "x": 30,
  "y": 20,
}
`;
